crypto/provider_core.c

```diff
@@ -513,33 +513,63 @@ static int create_provider_children(OSSL_PROVIDER *prov)
     return ret;
 }
 
-int ossl_provider_add_to_store(OSSL_PROVIDER *prov, int retain_fallbacks)
+int ossl_provider_add_to_store(OSSL_PROVIDER *prov, OSSL_PROVIDER **actualprov,
+                               int retain_fallbacks)
 {
-    struct provider_store_st *store = NULL;
-    int ret = 1;
+    struct provider_store_st *store;
+    int idx;
+    OSSL_PROVIDER tmpl = { 0, };
+    OSSL_PROVIDER *actualtmp = NULL;
 
     if ((store = get_provider_store(prov->libctx)) == NULL)
         return 0;
 
-
-    if (!ossl_provider_up_ref(prov)) {
-        ERR_raise(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE);
+    if (!CRYPTO_THREAD_write_lock(store->lock))
         return 0;
+
+    tmpl.name = (char *)prov->name;
+    idx = sk_OSSL_PROVIDER_find(store->providers, &tmpl);
+    if (idx == -1) {
+        if (sk_OSSL_PROVIDER_push(store->providers, prov) == 0)
+            goto err;
+        prov->store = store;
+        if (!create_provider_children(prov)) {
+            sk_OSSL_PROVIDER_delete_ptr(store->providers, prov);
+            goto err;
+        }
+        actualtmp = prov;
+        if (!retain_fallbacks)
+            store->use_fallbacks = 0;
+    } else {
+        actualtmp = sk_OSSL_PROVIDER_value(store->providers, idx);
     }
-    if (!CRYPTO_THREAD_write_lock(store->lock)
-            || sk_OSSL_PROVIDER_push(store->providers, prov) == 0) {
+
+    CRYPTO_THREAD_unlock(store->lock);
+
+    if (actualtmp != prov) {
+        /*
+         * The provider is already in the store. Probably two threads
+         * independently initialised their own provider objects with the same
+         * name and raced to put them in the store. This thread lost. We
+         * deactivate the one we just created and use the one that already
+         * exists instead.
+         */
+        ossl_provider_deactivate(prov);
         ossl_provider_free(prov);
-        ret = 0;
     }
-    prov->store = store;
-    if (!retain_fallbacks)
-        store->use_fallbacks = 0;
-    if (!create_provider_children(prov)) {
-        ret = 0;
+
+    if (actualprov != NULL) {
+        if (!ossl_provider_up_ref(actualtmp)) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE);
```

Comment:
Here the provider will be activated but the return value will be 0. Not sure if it is a problem or not.

Reply:
Maybe we should deactivate it in the `actualtmp == prov` case?

Reply:
At the moment, in the event of failure, this function returns 0 and has "done nothing". All callers then handle that appropriately (deactivate/free). That's fairly consistent with how we handle error elsewhere...i.e. do nothing and let the caller sort it out. So I think this is ok.

Reply:
Ah, ok. That's true that the `actualtmp == prov` case is fine. But what about the case where `actualtmp != prov`? In that case we return 0 but the original prov was deactivated and freed. But the caller does not know about that.

Reply:
Fixed